File: city-building-game.rb
class CityBuilderGame
  def initialize
    @money = 1000
    @population = 10
    @city_name = ""
    @buildings = []
  end

  def start
    puts "Welcome to the City Builder game!"
    puts "What's the name of your city?"
    @city_name = gets.chomp.capitalize
    puts "Your city name is #{@city_name}."
    puts "You have $#{@money} and a population of #{@population} people."
    puts "Let's start building your city!"

    loop do
      display_menu
      action = gets.chomp.downcase
      case action
      when "build"
        build_building
      when "destroy"
        destroy_building
      when "balance"
        display_balance
      when "population"
        display_population
      when "quit"
        puts "Thanks for playing!"
        break
      else
        puts "Invalid action."
      end
    end
  end

  def display_menu
    puts "------------------"
    puts "What do you want to do?"
    puts "1. Build a building"
    puts "2. Destroy a building"
    puts "3. Check balance"
    puts "4. Check population"
    puts "5. Quit"
    puts "------------------"
  end

  def build_building
    puts "What building do you want to build?"
    puts "1. House ($100, +1 population)"
    puts "2. Factory ($500, +$50/day)"
    choice = gets.chomp.to_i
    case choice
    when 1
      if @money >= 100
        @money -= 100
        @population += 1
        @buildings << "House"
        puts "You built a house! Your population is now #{@population}."
      else
        puts "You don't have enough money to build a house."
      end
    when 2
      if @money >= 500
        @money -= 500
        @buildings << "Factory"
        puts "You built a factory!"
      else
        puts "You don't have enough money to build a factory."
      end
    else
      puts "Invalid choice."
    end
  end

  def destroy_building
    puts "What building do you want to destroy?"
    @buildings.each_with_index { |building, index| puts "#{index+1}. #{building}" }
    choice = gets.chomp.to_i
    if choice.between?(1, @buildings.length)
      building = @buildings[choice-1]
      case building
      when "House"
        @population -= 1
        puts "You destroyed a house! Your population is now #{@population}."
      when "Factory"
        puts "You destroyed a factory."
      end
      @buildings.delete_at(choice-1)
    else
      puts "Invalid choice."
    end
  end

  def display_balance
    puts "Your balance is $#{@money}."
  end

  def display_population
    puts "Your population is #{@population}."
  end
end

game = CityBuilderGame.new
game.start

File: city-building-game.r
money <- 1000
population <- 10
city_name <- ""
buildings <- list()

display_menu <- function() {
  cat("------------------\n")
  cat("What do you want to do?\n")
  cat("1. Build a building\n")
  cat("2. Destroy a building\n")
  cat("3. Check balance\n")
  cat("4. Check population\n")
  cat("5. Quit\n")
  cat("------------------\n")
}

build_building <- function() {
  cat("What building do you want to build?\n")
  cat("1. House ($100, +1 population)\n")
  cat("2. Factory ($500, +$50/day)\n")
  choice <- as.numeric(readline())
  switch(choice,
         "1" = {
           if (money >= 100) {
             money <<- money - 100
             population <<- population + 1
             buildings <<- c(buildings, "House")
             cat("You built a house! Your population is now", population, ".\n")
           } else {
             cat("You don't have enough money to build a house.\n")
           }
         },
         "2" = {
           if (money >= 500) {
             money <<- money - 500
             buildings <<- c(buildings, "Factory")
             cat("You built a factory!\n")
           } else {
             cat("You don't have enough money to build a factory.\n")
           }
         },
         cat("Invalid choice.\n"))
}

destroy_building <- function() {
  cat("What building do you want to destroy?\n")
  for (i in seq_along(buildings)) {
    cat(i, ". ", buildings[[i]], "\n")
  }
  choice <- as.numeric(readline())
  if (choice %in% seq_along(buildings)) {
    building <- buildings[[choice]]
    switch(building,
           "House" = {
             population <<- population - 1
             cat("You destroyed a house! Your population is now", population, ".\n")
           },
           "Factory" = {
             cat("You destroyed a factory.\n")
           })
    buildings <<- buildings[-choice]
  } else {
    cat("Invalid choice.\n")
  }
}

display_balance <- function() {
  cat("Your balance is $", money, ".\n")
}

display_population <- function() {
  cat("Your population is ", population, ".\n")
}

start_game <- function() {
  cat("Welcome to the City Builder game!\n")
  cat("What's the name of your city?\n")
  city_name <<- readline()
  cat("Your city name is ", toupper(substr(city_name, 1, 1)), substr(city_name, 2), ".\n")
  cat("You have $", money, " and a population of ", population, " people.\n")
  cat("Let's start building your city!\n")
  repeat {
    display_menu()
    action <- readline()
    switch(tolower(action),
           "build" = {
             build_building()
           },
           "destroy" = {
             destroy_building()
           },
           "balance" = {
             display_balance()
           },
           "population" = {
             display_population()
           },
           "quit" = {
             cat("Thanks for playing!\n")
             break
           },
           cat("Invalid action.\n"))
  }
}

start_game()
